### PR TITLE
Event dispatching on disabled form controls

### PR DESCRIPTION
**Pull request by cdumez**
#### 9ed97a1ca58c955072d9cc1d1de82d4ab62e6dfb
<pre>
Event dispatching on disabled form controls
<a href="https://bugs.webkit.org/show_bug.cgi?id=251246">https://bugs.webkit.org/show_bug.cgi?id=251246</a>
rdar://104727624

Reviewed by Ryosuke Niwa.

Start dispatching events on disabled form controls behind a runtime flag.

This patch also makes &quot;click&quot;, &quot;mousedown&quot;, and &quot;mouseup&quot; events have their event
paths never include any disabled form controls or ancestors of disabled
form controls. This aligns with what Blink did in:
- <a href="https://chromium-review.googlesource.com/c/chromium/src/+/3929060">https://chromium-review.googlesource.com/c/chromium/src/+/3929060</a>
And is being discussed at:
- <a href="https://github.com/whatwg/html/issues/5886">https://github.com/whatwg/html/issues/5886</a>

Also make it so that fieldset elements can no longer be disabled, as per:
- <a href="https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#enabling-and-disabling-form-controls">https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#enabling-and-disabling-form-controls</a>:-the-disabled-attribute

This gets us closer to the specification and Chrome &amp; Firefox&apos;s behavior.

This is part of Interop 2023.

All new behavior is behind a runtime feature flag, currently off by default.

* LayoutTests/fast/css/pseudo-indeterminate-radio-buttons-basics-expected.html:
* LayoutTests/fast/forms/disabled-mousedown-event-expected.txt:
* LayoutTests/fast/forms/disabled-mousedown-event.html:
* LayoutTests/fast/forms/fieldset/fieldset-disabled-expected.txt:
* LayoutTests/fast/forms/fieldset/fieldset-disabled.html:
Resync these tests from Blink.

* LayoutTests/imported/w3c/web-platform-tests/dom/events/Event-dispatch-on-disabled-elements.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/disabled-elements/disabled-event-dispatch.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/disabled-elements/fieldset-event-propagation.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/disabled.html:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/enabled.html:
Resync these tests from WPT and rebaseline now that more checks are passing.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/disabled-elements/event-propagate-disabled.tentative-expected.txt:
Rebaseline now that more checks are passing. Note that some subtests are failing but I believe the test needs updating.
These subtests are failing in Blink as well:
<a href="https://chromium.googlesource.com/chromium/src.git/+/refs/heads/main/third_party/blink/web_tests/external/wpt/html/semantics/disabled-elements/event-propagate-disabled.tentative.html.ini">https://chromium.googlesource.com/chromium/src.git/+/refs/heads/main/third_party/blink/web_tests/external/wpt/html/semantics/disabled-elements/event-propagate-disabled.tentative.html.ini</a>

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::dispatchMouseEvent):
* Source/WebCore/dom/EventContext.cpp:
(WebCore::EventContext::handleLocalEvents const):
* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::EventDispatcher::dispatchEvent):
* Source/WebCore/dom/EventPath.cpp:
(WebCore::EventPath::adjustForDisabledFormControl):
* Source/WebCore/dom/EventPath.h:
* Source/WebCore/html/HTMLFieldSetElement.cpp:
(WebCore::HTMLFieldSetElement::isDisabledFormControl const):
* Source/WebCore/html/HTMLFieldSetElement.h:
* Source/WebCore/html/HTMLFormControlElement.h:

Canonical link: <a href="https://commits.webkit.org/264098@main">https://commits.webkit.org/264098@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5ca398753bf0441254f245c274ae7ce9227393a6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6662 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7060 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8253 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6941 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6661 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6826 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9840 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6780 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6098 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8369 "Built successfully") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6025 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13876 "7 flakes 1 missing results 114 failures") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5627 "Built successfully and passed tests") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6106 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8844 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/6250 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6596 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5404 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6799 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5992 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5953 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1579 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1580 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10164 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6982 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6365 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1707 "Passed tests") |

### CODE REVIEW

**Comment by nt1m** [comment]
Thanks for doing this! The radar link in the commit message needs updating fwiw

File: Source/WebCore/dom/EventPath.cpp [hunk 1]
```diff
@@ -177,6 +177,17 @@ void EventPath::setRelatedTarget(Node& origin, Node& relatedNode)
     }
 }
 
+void EventPath::adjustForDisabledFormControl()
+{
+    for (unsigned i = 0; i < m_path.size(); ++i) {
```
Comment (rniwa) [inline Source/WebCore/dom/EventPath.cpp:182]: Where in the spec does it say to do this?

Comment (cdumez) [reply Source/WebCore/dom/EventPath.cpp:182]: As far as I know it doesn't currently but this is Firefox behavior. This is also what Blink did more recently to align:
- https://chromium-review.googlesource.com/c/chromium/src/+/3929060

This is what's being discussed here (seemed to be the most popular behavior):
- https://github.com/whatwg/html/issues/5886

With us doing this too, all 3 engines should agree on this behavior now.

Comment (cdumez) [reply Source/WebCore/dom/EventPath.cpp:182]: I added this information to the commit log.

**Review by rniwa — APPROVED** [review]

**Pull request merged**